semantic-codegen/src/AST/GenerateSyntax.hs

```diff
@@ -71,34 +72,55 @@ syntaxDatatype language allSymbols datatype = skipDefined $ do
     ProductType (DatatypeName datatypeName) named children fields -> do
       con <- ctorForProductType datatypeName typeParameterName children fields
-      result <- symbolMatchingInstance allSymbols name named datatypeName
-      pure $ generatedDatatype name [con] typeParameterName:result
+      symbolMatchingInstance <- symbolMatchingInstance allSymbols name named datatypeName
+      traversalInstances <- makeTraversalInstances (conT name)
+      pure
+        (  generatedDatatype name [con] typeParameterName
+        :  symbolMatchingInstance
+        <> traversalInstances)
       -- Anonymous leaf types are defined as synonyms for the `Token` datatype
     LeafType (DatatypeName datatypeName) Anonymous -> do
       tsSymbol <- runIO $ withCStringLen datatypeName (\(s, len) -> TS.ts_language_symbol_for_name language s len False)
       pure [ TySynD name [] (ConT ''Token `AppT` LitT (StrTyLit datatypeName) `AppT` LitT (NumTyLit (fromIntegral tsSymbol))) ]
     LeafType (DatatypeName datatypeName) Named -> do
       con <- ctorForLeafType (DatatypeName datatypeName) typeParameterName
-      result <- symbolMatchingInstance allSymbols name Named datatypeName
-      pure $ generatedDatatype name [con] typeParameterName:result
+      symbolMatchingInstance <- symbolMatchingInstance allSymbols name Named datatypeName
+      traversalInstances <- makeTraversalInstances (conT name)
+      pure
+        (  generatedDatatype name [con] typeParameterName
+        :  symbolMatchingInstance
+        <> traversalInstances)
   where
     -- Skip generating datatypes that have already been defined (overridden) in the module where the splice is running.
     skipDefined m = do
       isLocal <- lookupTypeName nameStr >>= maybe (pure False) isLocalName
       if isLocal then pure [] else m
     name = mkName nameStr
     nameStr = toNameString (datatypeNameStatus datatype) (getDatatypeName (AST.Deserialize.datatypeName datatype))
-    deriveStockClause = DerivClause (Just StockStrategy) [ ConT ''Eq, ConT ''Ord, ConT ''Show, ConT ''Generic, ConT ''Foldable, ConT ''Functor, ConT ''Traversable, ConT ''Generic1]
-    deriveAnyClassClause = DerivClause (Just AnyclassStrategy) [ConT ''TS.Unmarshal]
+    deriveStockClause = DerivClause (Just StockStrategy) [ ConT ''Eq, ConT ''Ord, ConT ''Show, ConT ''Generic, ConT ''Generic1]
+    deriveAnyClassClause = DerivClause (Just AnyclassStrategy) [ConT ''TS.Unmarshal, ConT ''Traversable1 `AppT` VarT (mkName "someConstraint")]
     deriveGN = DerivClause (Just NewtypeStrategy) [ConT ''TS.SymbolMatching]
     generatedDatatype name cons typeParameterName = DataD [] name [PlainTV typeParameterName] Nothing cons [deriveStockClause, deriveAnyClassClause]
 
 
+makeTraversalInstances :: TypeQ -> Q [Dec]
+makeTraversalInstances ty =
+  [d|
+  instance Foldable $ty where
+    foldMap = foldMapDefault1
+  instance Functor $ty where
+    fmap = fmapDefault1
+  instance Traversable $ty where
+    traverse = traverseDefault1
```

Comment:
These instances compile significantly faster than the derived ones did, which more than makes up for the addition of the new `Traversable1` instance.